### PR TITLE
[SPARK-44591][CONNECT][WEBUI] Use jobTags in SparkListenerSQLExecutionStart to link SQL Execution IDs for Spark UI Connect page

### PR DESCRIPTION
**Pull request by jasonli-db**
### What changes were proposed in this pull request?
Use jobTags in SparkListenerSQLExecutionStart to get the corresponding SQL Execution for Spark Connect request rather than SparkListenerJobStart.props.getProperty(SQLExecution.EXECUTION_ID_KEY), which won't work when the SQL Execution does not trigger a job.

### Why are the changes needed?
This change handles cases where a SQL Execution doesn't trigger a job and we can't retrieve the SQL Execution ID from SparkListenerJobStart

### How was this patch tested?
Update unit test + local manual testing

### CODE REVIEW

**Review by juliuszsompolski — APPROVED** [review]
LGTM,
should follow the other Spark Connect UI PRs into branch-3.5
cc @gengliangwang @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala [hunk 1]
```diff
@@ -128,6 +126,43 @@ private[connect] class SparkConnectServerListener(
     }
   }
 
+  def onSQLExecutionStart(e: SparkListenerSQLExecutionStart): Unit = {
+    val executeJobTagOpt = e.jobTags.find {
```
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:130]: QQ: why not using executionId from SparkListenerSQLExecutionStart?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala [hunk 2]
```diff
@@ -128,6 +126,43 @@ private[connect] class SparkConnectServerListener(
...
+    }
+    val executeJobTag = executeJobTagOpt.get
+    val exec = executionList.get(executeJobTag)
+    if (exec.nonEmpty) {
+      exec.foreach { exec =>
+        exec.sqlExecId += e.executionId.toString
+        updateLiveStore(exec)
+      }
+    } else {
+      // This block guards against potential event re-ordering.
```
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:145]: Can you explain the event re-ordering here?

Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:145]: It's to guard against SQLStart and JobStart events potentially coming after an OperationClosed event. It's mirrors the one in onJobStart (which was included in the Thrift Server listener as well). See the comment here: https://github.com/apache/spark/pull/42244/files#diff-ffa4c6fbd382a52c08a0f10b7fbbd6c2d09de322f7e3145ecba464f0cda5843dR89-R92

Comment (gengliangwang) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:145]: Let's extend the comment in the code

Comment (gengliangwang) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:145]: BTW, how often can the event reorder happen? It looks odd to update the data in KVStore.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master/branch-3.5

**Pull request closed without merging**